compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs

```diff
@@ -211,35 +211,11 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
 
             // Suggest removing a `&mut` from the use of a mutable reference.
             PlaceRef { local, projection: [] }
-                if {
-                    self.body
-                        .local_decls
-                        .get(local)
-                        .map(|local_decl| {
-                            if let Some(box LocalInfo::User(ClearCrossCrate::Set(
-                                mir::BindingForm::ImplicitSelf(kind),
-                            ))) = local_decl.local_info
-                            {
-                                // Check if the user variable is a `&mut self` and we can therefore
-                                // suggest removing the `&mut`.
-                                //
-                                // Deliberately fall into this case for all implicit self types,
-                                // so that we don't fall in to the next case with them.
-                                kind == mir::ImplicitSelfKind::MutRef
-                            } else if Some(kw::SelfLower) == self.local_names[local] {
-                                // Otherwise, check if the name is the self kewyord - in which case
-                                // we have an explicit self. Do the same thing in this case and check
-                                // for a `self: &mut Self` to suggest removing the `&mut`.
-                                if let ty::Ref(_, _, hir::Mutability::Mut) = local_decl.ty.kind() {
-                                    true
-                                } else {
-                                    false
-                                }
-                            } else {
-                                false
-                            }
-                        })
-                        .unwrap_or(false)
+                if match self.body.local_decls.get(local) {
```

Comment:
Should we refactor this into another function? https://github.com/rust-lang/rust/blob/81e02708f1f4760244756548981277d5199baa9a/compiler/rustc_middle/src/mir/mod.rs#L937
as

```rust
                if self.body.local_decls
                    .get(local)
                    .map_or(false, |local_decl| local_decl.mut_borrow_of_mutable_ref(self.local_names[local])) => {
```
But I wonder if it will even failed to get `local`? The next block just use `local_decls[local]` but not sure if it will work here.
```rust
                if self.body.local_decls[local]
                    .mut_borrow_of_mutable_ref(self.local_names[local]) => {
```

Reply:
I don't get it. Did you mean to make a method on LocalDecls?

Reply:
Yeah

Reply:
I don't think there are code elsewhere that need to reuse the logic.

Reply:
Yes, that's true but I don't know if putting it as another function on itself would be good. Maybe put is as an implementation under `MirBorrowckCtxt` which is the same thing?

Reply:
Honestly I prefer free function as it since it don't need other states than passed arguments.
But I don't have strong opinion, let's let it for Esteban or David.

Reply:
I don't know if it is worth extracting this into a method, but I would prefer something like this than a `match` here:

```rust
self.body.local_decls
    .get(local)
    .map(|l| mut_borrow_of_mutable_ref(l, self.local_names[local])
    .unwrap_or(false)
```

Reply:
Done.